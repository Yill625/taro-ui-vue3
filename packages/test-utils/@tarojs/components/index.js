export const Text = 'text'
export const Image = 'image'
export const Icon = 'icon'
export const ScrollView = 'scroll-view'
export const View = 'view'
export const RichText = 'rich-text'
export const Button = 'button'
export const CheckboxGroup = 'checkbox-group'
export const Editor = 'editor'
export const Form = 'form'
export const Label = 'label'
export const PickerView = 'picker-view'
export const PickerViewColumn = 'picker-view-column'
export const CoverImage = 'cover-image'
export const CoverView = 'cover-view'
export const MoveableArea = 'moveable-area'
export const MoveableView = 'moveable-view'
export const Swiper = 'swiper'
export const FunctionalPageNavigator = 'functional-page-navigator'
export const Navigator = 'navigator'
export const Audio = 'audio'
export const Camera = 'camera'
export const LivePlayer = 'live-player'
export const Map = 'map'
export const Ad = 'ad'
export const OfficialAccount = 'official-account'
export const OpenData = 'open-data'
export const WebView = 'web-view'
export const NavigationBar = 'navigation-bar'
export const Block = 'block'
export const Canvas = 'canvas'
export const Checkbox = 'checkbox'
export const Progress = 'progress'
export const RadioGroup = 'radio-group'
export const Radio = 'radio'
export const SwiperItem = 'swiper-item'
export const Video = 'video'
export const Input = 'input'
export const Textarea = 'textarea'
export const Picker = 'picker'
export const Switch = 'switch'
export const Slider = 'slider'