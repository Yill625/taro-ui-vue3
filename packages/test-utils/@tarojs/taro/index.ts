
const ENV_TYPE = {
  WEAPP: 'WEAPP',
  WEB: 'WEB',
  RN: 'RN',
  SWAN: 'SWAN',
  ALIPAY: 'ALIPAY',
  TT: 'TT',
  QQ: 'QQ',
  JD: 'JD',
  QUICKAPP: 'QUICKAPP'
}

const envs = {
  "weapp": ENV_TYPE.WEAPP,
  "swan": ENV_TYPE.SWAN,
  "alipay": ENV_TYPE.ALIPAY,
  "h5": ENV_TYPE.WEB,
  "tt": ENV_TYPE.TT,
  "quickapp": ENV_TYPE.QUICKAPP,
  "qq": ENV_TYPE.QQ,
  "jd": ENV_TYPE.JD,
}

export const getEnvMockFn = () => {
  if (!process.env.TARO_ENV) {
    process.env.TARO_ENV = 'h5'
  }
  return envs[process.env.TARO_ENV]
}

export const pxTransformMockFn = (size, designWidth = 750) => {
  if (!process.env.TARO_ENV) {
    process.env.TARO_ENV = 'h5'
  }

  const deviceRatio = {
    640: 2.34 / 2,
    750: 1,
    828: 1.81 / 2
  }

  if (!size) return ''

  return process.env.TARO_ENV === 'h5'
    ? Math.ceil(parseInt(`${size}`, 10) / 40 * 640 / designWidth * 10000) / 10000 + 'rem'
    : parseInt(size, 10) * deviceRatio[designWidth] + 'rpx'
}

export const querySelectorMockFn = (
  rect0 = { width: 30, left: 30, height: 30, right: 30 }
) => (): any => {
  const query = {
    exec: (cb) => { cb([rect0]) }
  }
  query['select'] = jest.fn().mockReturnValue({
    boundingClientRect: jest.fn().mockReturnValue(query)
  })

  return query
}

export const chooseImageMockFn = () => {
  return new Promise<any>((resolve) => {
    resolve({
      tempFilePaths: ['/path/to/file'],
      tempFiles: [{ path: '/path/to/file', size: 256 }],
      errMsg: ''
    })
  })
}

document.querySelector = jest.fn((sel) => {
  return {
    getBoundingClientRect: jest.fn().mockReturnValueOnce({ width: 30 })
  }
})