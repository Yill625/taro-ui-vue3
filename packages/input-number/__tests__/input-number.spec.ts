import {
  genMountFn,
  testPropClassAndStyle
} from '@taro-ui-vue3/test-utils/helper'
import { pxTransformMockFn } from '@taro-ui-vue3/test-utils/@tarojs/taro'
import AtInputNumber from '../index.vue'
import { ref } from 'vue'

const mountFn = genMountFn(AtInputNumber)

describe('AtInputNumber', () => {
  testPropClassAndStyle(mountFn)

  it('should render all nodes and match snapshot', () => {
    const wrapper = mountFn()
    expect(wrapper.element).toMatchSnapshot()
  })

  it.each([
    "number",
    "digit"
  ])('should render prop type -- %s', (typeOption) => {
    const wrapper = mountFn({ type: typeOption })
    expect(
      wrapper
        .find('.at-input-number__input')
        .attributes('type')
    ).toBe(typeOption)
  })

  it('should render prop width', () => {
    const wrapper = mountFn({ width: 200 })
    expect(
      wrapper
        .find('.at-input-number__input')
        .attributes('style')
    ).toBe(`width: ${pxTransformMockFn(200)};`)
  })

  it.each([
    "normal",
    "large"
  ])('should render prop size -- %s', (sizeOption) => {
    const wrapper = mountFn({ size: sizeOption })
    expect(
      wrapper
        .find('.at-input-number--lg')
        .exists()
    ).toBe(sizeOption === 'large')
  })

  it('should render prop disabled', () => {
    const wrapper = mountFn({ disabled: true })
    expect(
      wrapper
        .findAll('.at-input-number--disabled')
        .length
    ).toBe(2)
  })

  it('should disable minus button if input value <= min', () => {
    const wrapper = mountFn({
      min: 5,
      modelValue: 1
    })

    expect(
      wrapper
        .find('.at-input-number__btn:first-child')
        .classes('at-input-number--disabled')
    ).toBe(true)

    expect(
      wrapper
        .find('.at-input-number__btn:last-child')
        .classes('at-input-number--disabled')
    ).toBe(false)
  })

  it('should disable plus button if input value >= max', () => {
    const wrapper = mountFn({
      max: 100,
      modelValue: 110
    })

    expect(
      wrapper
        .find('.at-input-number__btn:last-child')
        .classes('at-input-number--disabled')
    ).toBe(true)

    expect(
      wrapper
        .find('.at-input-number__btn:first-child')
        .classes('at-input-number--disabled')
    ).toBe(false)
  })

  it('should increase value at given step', async () => {
    const original = ref(5)
    const wrapper = mountFn({
      modelValue: original.value,
      step: 5,
      'onUpdate:modelValue': (e) => original.value = e
    })

    await wrapper
      .find('.at-input-number__btn:last-child')
      .trigger('tap')

    expect(original.value).toBe(10)
  })

  it('should decrease value at given step', async () => {
    const original = ref(10)
    const wrapper = mountFn({
      modelValue: original.value,
      step: 5,
      'onUpdate:modelValue': (e) => original.value = e
    })

    await wrapper
      .find('.at-input-number__btn:first-child')
      .trigger('tap')

    expect(original.value).toBe(5)
  })
})

describe('AtNumberInput events', () => {

  it('should emit `update:modelValue`', async () => {
    const onUpdateValue = jest.fn()
    const wrapper = mountFn({
      modelValue: 2,
      'onUpdate:modelValue': onUpdateValue
    })

    await wrapper
      .find('.at-input-number .at-input-number__input')
      .trigger('input')

    expect(onUpdateValue).toBeCalled()
    expect(wrapper.emitted()).toHaveProperty('update:modelValue')
  })

  it('should not emit `update:modelValue` when disabled', async () => {
    const onUpdateValue = jest.fn()
    const wrapper = mountFn({
      modelValue: 2,
      disabled: true,
      'onUpdate:modelValue': onUpdateValue
    })

    await wrapper
      .find('.at-input-number .at-input-number__input')
      .trigger('input')

    expect(wrapper.emitted()).not.toHaveProperty('update:modelValue')
    expect(onUpdateValue).not.toBeCalled()
  })

  it('should emit blur', async () => {
    const onBlur = jest.fn()
    const wrapper = mountFn({ modelValue: 2, onBlur })

    await wrapper
      .find('.at-input-number .at-input-number__input')
      .trigger('blur')

    expect(wrapper.emitted()).toHaveProperty('blur')
    expect(onBlur).toBeCalled()
  })

  it('should emit error-input when value after minus or plus is below min or over max', async () => {
    const onErrorInput = jest.fn()
    const wrapper = mountFn({
      modelValue: 2,
      step: 50,
      onErrorInput
    })

    await wrapper
      .find('.at-input-number__btn:first-child')
      .trigger('tap')

    await wrapper.setProps({ modelValue: 90 })
    await wrapper
      .find('.at-input-number__btn:last-child')
      .trigger('tap')

    expect(wrapper.emitted()).toHaveProperty('error-input')

    expect(onErrorInput.mock.calls.length).toBe(2)
    expect(onErrorInput.mock.calls[0][0]).toEqual({
      type: 'LOW',
      errorValue: 0
    })
    expect(onErrorInput.mock.calls[1][0]).toEqual({
      type: 'OVER',
      errorValue: 100
    })
  })

  it('should emit error-input when initial value is below min or over max', async () => {
    const onErrorInput = jest.fn()
    const wrapper = mountFn({
      modelValue: -10,
      onErrorInput
    })

    await wrapper
      .find('.at-input-number__btn:first-child')
      .trigger('tap')

    await wrapper.setProps({ modelValue: 110 })
    await wrapper
      .find('.at-input-number__btn:last-child')
      .trigger('tap')

    expect(wrapper.emitted()).toHaveProperty('error-input')
    expect(onErrorInput.mock.calls.length).toBe(4)
    expect(onErrorInput.mock.calls[0][0]).toEqual({
      type: 'LOW',
      errorValue: 0
    })
    expect(onErrorInput.mock.calls[1][0]).toEqual({
      type: 'LOW',
      errorValue: -1
    })
    expect(onErrorInput.mock.calls[2][0]).toEqual({
      type: 'OVER',
      errorValue: 100
    })
    expect(onErrorInput.mock.calls[3][0]).toEqual({
      type: 'OVER',
      errorValue: 101
    })
  })

  it('should emit error-input when disabled', async () => {
    const onErrorInput = jest.fn()
    const wrapper = mountFn({
      modelValue: 2,
      step: 1,
      disabled: true,
      onErrorInput
    })

    await wrapper
      .find('.at-input-number__btn:first-child')
      .trigger('tap')

    expect(wrapper.emitted()).toHaveProperty('error-input')
    expect(onErrorInput).toHaveBeenCalledWith({
      type: 'DISABLED',
      errorValue: 1
    })
  })
})
