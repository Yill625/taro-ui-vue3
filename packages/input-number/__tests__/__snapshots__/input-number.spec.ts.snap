// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtInputNumber should render all nodes and match snapshot 1`] = `
<view
  class="at-input-number"
>
  <view
    class="at-input-number__btn"
  >
    <text
      class="at-icon at-icon-subtract at-input-number__btn-subtract"
    />
  </view>
  <input
    class="at-input-number__input"
    style="width: 2.56rem;"
    type="number"
  />
  <view
    class="at-input-number__btn"
  >
    <text
      class="at-icon at-icon-add at-input-number__btn-add"
    />
  </view>
</view>
`;
